Update keyboard controls help in the game and the editor and generate it semi-automatically

tothxa opened the pull request:
**Type of change**

Bugfix / Refactoring

**Issues closed**

fixes #5716

Last fastplace slot can actually be used, not only defined.

**New behaviour**

Keyboard shortcuts help texts in the Encyclopedia and the in-editor help are generated from the same definition that is used for the keyboard options window. This should help make sure that new hotkeys will not be left out in the future.

Fastplace shortcuts for the current player/tribe are now also included in the Encyclopedia.

The hotkeys for debug features (debug console and cheat mode) were made configurable, and a hotkey was added for upload map in the editor.

**Possible regressions**

Anything related to keyboard shortcuts, because the enum class was reordered heavily.

tothxa committed:
Update hotkey references in editor help
fixes #5716

tothxa commented:
This is a draft because although it already fixes the bug itself by updating the existing hotkeys in the help, it does not add the new ones yet. And I believe there's room for more cleanup too.

frankystone commented:
The Key combination for opening the debug console is missing in the editor: SHIFT+CTRL+SPACE

tothxa committed:
add some common hotkeys to editor
fix translator comments
update help text on replay files

tothxa committed:
Merge branch 'master' into fix-hotkey-help

tothxa commented:
I've added the help for the debug console.

While editing the common helptexts file, I've noticed there was a text for the old 2 part replay format, so I updated that too to reflect the new single file format.

tothxa committed:
fix missing "

tothxa committed:
replay helptext improvement

tothxa committed:
Merge branch 'master' into fix-hotkey-help

tothxa committed:
Merge branch 'master' into fix-hotkey-help

tothxa committed:
add missing hotkeys to game help
add tip to set mousewheel options for touchpad

tothxa committed:
groundwork for autogenerating shortcut help

tothxa committed:
Merge branch 'master' into fix-hotkey-help

tothxa committed:
make it compile and add tests to in-game/in-editor help

tothxa committed:
add hotkey for upload map

tothxa committed:
add as_definition_line() formatting function

tothxa committed:
2 files were automatically formatted.

tothxa committed:
fix codecheck

tothxa committed:
try to simplify iterating over keyboard shortcuts

tothxa committed:
'src/wlapplication_options.cc' was automatically formatted.

tothxa committed:
* simplify all iterations over keyboard shortcuts
 * allow using message- and ship stats window shortcut names with or without prefix

tothxa committed:
make debug console and cheat mode hotkeys configurable

tothxa committed:
editor hotkeys help mostly done
 * reorder shortcuts enum for inclusion in help
 * include the localised ":" in as_definition_line()

tothxa committed:
3 files were automatically formatted.

tothxa committed:
complete both in-game and editor hotkeys help generator code

tothxa committed:
'src/wlapplication_options.cc' was automatically formatted.

tothxa committed:
fix translators comments clobbered by clang-format

tothxa commented:
This is done functionally, but some descriptions may still need changing to make them more suitable for the help too.

tothxa committed:
Merge branch 'master' into fix-hotkey-help

tothxa committed:
add fastplace shortcuts to encyclopedia

tothxa committed:
2 files were automatically formatted.

tothxa committed:
remove leftover include

tothxa committed:
fix codecheck and off-by-one problem

tothxa committed:
clang-tidy

tothxa committed:
clang-tidy

tothxa committed:
Merge branch 'master' into fix-hotkey-help

tothxa committed:
Merge branch 'master' into fix-hotkey-help

tothxa committed:
simplify help generation by implementing aliases for common hotkeys

tothxa committed:
simplify special sections too

tothxa committed:
2 files were automatically formatted.

tothxa committed:
fix clang-tidy and hide implementation helper functions

tothxa committed:
'src/wlapplication_options.cc' was automatically formatted.

tothxa committed:
improve check

tothxa commented:
@Noordfrees May I convert `KeyboardShortcutInfo` to store the English descriptions as `std::string`, instead of the current `std::function`? These would then be initialised using `gettext_noop()`, and translated in `to_string()`.

Either here or in a separate PR, because this one already looks much bigger than it actually is.

tothxa committed:
convert shortcuts_ to store descnames as strings instead of functions

tothxa committed:
'src/wlapplication_options.cc' was automatically formatted.

tothxa commented:
@frankystone Could you please test this again?

frankystone commented:
Some brief testing and observations, but i am a mouse jockey and those shortcuts is a closed book for me :smile:  :

Some shortcut entries in the Editors encyclopedia are only valid for a game. E.g. `,`, `.` or `STRG+SHIFT+BACKSPACE` (cheating). In a game help i found `Entf` (Delete). I think those shortcuts shouldn't be mixed. The editor help should only contain valid shortcuts for the editor and the game help should show only valid shortcuts for a game

Otherwise this works as expected for me.

tothxa commented:
Thank you for testing.

> Some shortcut entries in the Editors encyclopedia are only valid for a game. E.g. `,`, `.` or `STRG+SHIFT+BACKSPACE` (cheating). In a game help i found `Entf` (Delete). I think those shortcuts shouldn't be mixed. The editor help should only contain valid shortcuts for the editor and the game help should show only valid shortcuts for a game

The new  help generator code uses the same groups as the shortcut settings in the Options menu, so yes, only valid ones should be listed. And actually most of your examples do work in both:

 - `,` and `.` work in the editor: if you jump around by clicking on the minimap, you can then backtrack with `,` and then forward again with `.`.
 - I was surprised you didn't expect the `Delete` key in the in-game menu, because it's used by the messages window. So I searched for it, and it turns out that the actual configurable hotkey is only referenced by that and the save/load game windows, but nothing in the editor code. The `Delete` key itself is hardcoded though in some basic UI elements for text editing.

As for cheating mode, that's a newly configurable hotkey in this PR. I included it in common, because it's used in common code (`interactive_base.cc`) right next to the debug console hotkey, so even if it has no real effect in the editor, it could mask other editor hotkeys if allowed to be set to the same combination.

tothxa committed:
Merge branch 'master' into fix-hotkey-help

tothxa committed:
Merge branch 'master' into fix-hotkey-help

tothxa committed:
Merge branch 'master' into fix-hotkey-help

Noordfrees reviewed (commented):
Looks cool, not tested yet, some comments inline

tothxa committed:
Merge branch 'master' into fix-hotkey-help

tothxa committed:
apply suggestion

Co-authored-by: Benedikt Straub <48293446+Noordfrees@users.noreply.github.com>

tothxa committed:
'src/wlapplication_options.cc' was automatically formatted.

tothxa committed:
address some review comments

tothxa committed:
Merge branch 'master' into fix-hotkey-help

tothxa committed:
Merge branch 'master' into fix-hotkey-help

tothxa committed:
address remaining review comments

tothxa committed:
Merge branch 'master' into fix-hotkey-help

Noordfrees committed:
Merge branch 'master' into fix-hotkey-help

Noordfrees commented:
Hm, tested it again and works fine, but I'm still a bit unhappy about the Console and Cheat hotkeys. Please do hide them also in the options in release builds. Other than that LGTM.

tothxa commented:
> Hm, tested it again and works fine, but I'm still a bit unhappy about the Console and Cheat hotkeys. Please do hide them also in the options in release builds.

But then how can the user resolve a conflict with it? (ie. free ctrl+shift+space to be able to set it to something else)

tothxa committed:
hide debug hotkeys in release builds and silently clear them on conflict

tothxa committed:
Merge branch 'master' into fix-hotkey-help

tothxa commented:
@Noordfrees I hope this way is acceptable to you. But I'm doing it against my will only to get the fixes merged at last. :-/

(I forgot that `space` and `backspace` cannot actually be assigned, so testing this needs changing them first in a debug build)

Noordfrees reviewed (approved):
Thanks, looking good :)

Noordfrees committed:
Merge branch 'master' into fix-hotkey-help
